src/QtUtils/FutureWatcher.cpp

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "QtUtils/FutureWatcher.h"
+
+#include <absl/synchronization/mutex.h>
+
+#include <QDebug>
+#include <QEventLoop>
+#include <QObject>
+#include <QPointer>
+#include <QTimer>
+#include <chrono>
+
+namespace orbit_qt_utils {
+
+FutureWatcher::FutureWatcher(QObject* parent) : QObject{parent} {}
+
+void FutureWatcher::Abort() { emit AbortRequested(); }
+
+FutureWatcher::Reason FutureWatcher::WaitFor(
+    const orbit_base::Future<void>& future,
+    std::optional<std::chrono::milliseconds> timeout) const {
+  if (!future.IsValid() || future.IsFinished()) return Reason::kFutureCompleted;
+
+  QTimer timer{};
+  timer.setSingleShot(true);
+
+  QEventLoop loop{};
+  QObject::connect(this, &FutureWatcher::AbortRequested, &loop, &QEventLoop::quit);
+  QObject::connect(&timer, &QTimer::timeout, &loop, &QEventLoop::quit);
+
+  if (timeout.has_value()) {
+    timer.start(timeout.value());
+  }
+
+  future.RegisterContinuation(
+      [loop = QPointer{&loop}] { QMetaObject::invokeMethod(loop, &QEventLoop::quit); });
+
+  // Ensure that no race condition occured. The futured could have completed while we were setting
+  // the continuation.
+  if (future.IsFinished()) return Reason::kFutureCompleted;
+
+  loop.exec();
+
+  if (future.IsFinished()) return Reason::kFutureCompleted;
+
+  constexpr int kInvalidRemainingTime = -1;
+  if (timeout && timer.remainingTime() == kInvalidRemainingTime) return Reason::kTimeout;
```

Comment:
Above you used `timeout.has_value()`. Please be consistent

Reply:
Done.